### ISSUE TITLE
Fatal IO error 11

### ISSUE THREAD

**Issue by pychess** [pychess/pychess#821]
Original [issue 821](https://code.google.com/p/pychess/issues/detail?id=821) reported by [gbtami](https://code.google.com/u/gbtami/) 2013-10-25

Played some game on FICS. After my partner (lohem) resigned, I got:

tamas@tami:~$ LANG=C p

*\* (process:7877): WARNING **: Trying to register gtype 'GMountMountFlags' as enum when in fact it is of type 'GFlags'

*\* (process:7877): WARNING **: Trying to register gtype 'GDriveStartFlags' as enum when in fact it is of type 'GFlags'

*\* (process:7877): WARNING *_: Trying to register gtype 'GSocketMsgFlags' as enum when in fact it is of type 'GFlags'
*_ Message: pygobject_register_sinkfunc is deprecated (GstObject)
pychess: Fatal IO error 11 (Resource temporarily unavailable) on X server :0.0.

The tail of log:
14:44:10 ('Ping', '14:10:6.839') Debug: 64 bytes from fics.freechess.org (69.36.243.188): icmp_seq=4 ttl=51 time=211 ms
14:44:10 ('GuestVHGG', 'raw') Debug: 
lohem is no longer available for matches.
fics% 
<wd> lohem
fics% 
14:44:10 ('GuestVHGG', 'nonmatched') Debug: 
14:44:10 ('GuestVHGG', 'lines') Debug: lohem is no longer available for matches.
14:44:10 ('GuestVHGG', 'on_player_unavailable') Debug: lohem is no longer available for matches.
14:44:10 ('GuestVHGG', 'lines') Debug: 
14:44:10 ('GuestVHGG', 'nonmatched') Debug: 
14:44:10 ('GuestVHGG', 'lines') Debug: <wd> lohem

**Comment by pychess** [comment]
Original [comment 1.](https://code.google.com/p/pychess/issues/detail?id=821#c1) posted by [mattgatto](https://code.google.com/u/mattgatto/) 2013-10-25

Is it reproduceable? I haven't seen this. Maybe it's because I haven't been winning enough games ;)

**Comment by pychess** [comment]
Original [comment 2.](https://code.google.com/p/pychess/issues/detail?id=821#c2) posted by [gbtami](https://code.google.com/u/gbtami/) 2013-10-25

It's in Debian bugtracker too: http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=659377
It happens once in a blue moon, at least for me.

**Comment by pychess** [comment]
Original [comment 3.](https://code.google.com/p/pychess/issues/detail?id=821#c3) posted by [gbtami](https://code.google.com/u/gbtami/) 2014-02-18

It happened again today. After my FICS game finished, my partner sent me a "thank you" chat message. Right after that, pychess quit.
- attached [tail.log](https://code.google.com/p/pychess/issues/detail?id=821#c3)

**Comment by pychess** [comment]
Original [comment 4.](https://code.google.com/p/pychess/issues/detail?id=821#c4) posted by [mattgatto](https://code.google.com/u/mattgatto/) 2014-02-21

Apparently these X server messages are usually the result of threading errors where a thread is trying to manipulate the GUI without having the glock/gdk-lock (via threads_enter()). It seems like the common "thread" between these two crashes is that a user you played a game against disconnected. The only areas of code that receive callbacks from FICSPlayerExited are the Players List tab and the Chat window. In the Players List tab, the callback is locked, so that's not the culprit:

class PlayerTabSection (ParrentListSection):
[...]
    @glock.glocked
    def onPlayerAdded (self, players, player):

But the callback in the ChatWindow.py is not locked:

```
def addRow (self, id, text, type):
    if id in self.id2iter: return
    iter = self.props.model.append([id, text, type])
    self.id2iter[id] = iter
    self.idSet.add(id)
```

[...]
        def addPlayer (players, player):
            self.playersList.addRow(self.compileId(player.name, TYPE_PERSONAL),
                player.name + player.display_titles(), TYPE_PERSONAL)
            return False
        self.connection.players.connect("FICSPlayerEntered", addPlayer)
        def removePlayer (players, player):
            self.playersList.removeRow(self.compileId(player.name, TYPE_PERSONAL))
            return False
        self.connection.players.connect("FICSPlayerExited", removePlayer)

And I verified this is the case (i.e., that addRow() can get called from a non-Main thread without the gdk lock) via an Eclipse run. This is shown in the attached screenshot.

Fix coming...
- attached [pychess-GUI-update-without-gdk-lock.png](https://code.google.com/p/pychess/issues/detail?id=821#c4)

**Comment by pychess** [comment]
Original [comment 5.](https://code.google.com/p/pychess/issues/detail?id=821#c5) posted by [mattgatto](https://code.google.com/u/mattgatto/) 2014-02-21

This issue was closed by revision eb97966cef90.
- changed status to: Fixed

**Comment by pychess** [comment]
Original [comment 6.](https://code.google.com/p/pychess/issues/detail?id=821#c6) posted by [gbtami](https://code.google.com/u/gbtami/) 2014-02-22

It happened again today.
- attached [tail.log](https://code.google.com/p/pychess/issues/detail?id=821#c6)
- changed status to: Reopened

**Comment by pychess** [comment]
Original [comment 7.](https://code.google.com/p/pychess/issues/detail?id=821#c7) posted by [mattgatto](https://code.google.com/u/mattgatto/) 2014-02-23

This issue was closed by revision df77ba0efc4e.
- changed status to: Fixed

**Comment by pychess** [comment]
Original [comment 8.](https://code.google.com/p/pychess/issues/detail?id=821#c8) posted by [mattgatto](https://code.google.com/u/mattgatto/) 2014-02-23

I think the crash from your comment comment 6. was different than the previous crashes in this issue report. The previous logs ended with:
00:21:18 ('GuestTSHD', 'lines') DEBUG: <wd> Preserve

... But comment comment 6. ended with:
10:37:04 Default DEBUG: BM.onPlayGameCreated: Creating: lmftel (1408) gbtami (1413) rated lightning 2 1
{Game 224 (lmftel vs. gbtami) Creating rated lightning match.}
10:37:04 ('Ping', '09:02:34.353') DEBUG: 64 bytes from fics.freechess.org (69.36.243.188): icmp_seq=2422 ttl=52 time=230 ms

... And since gamewidget creation is all done using GtkWorker and Publisher which have gdk locking built-in, it must have been the pinger updating the ping time in the FICS Lounge without the gdk lock.

**Comment by pychess** [comment]
Original [comment 9.](https://code.google.com/p/pychess/issues/detail?id=821#c9) posted by [gbtami](https://code.google.com/u/gbtami/) 2014-02-28

It happened again.
- attached [2014-02-28_09-57-26.log.bz2](https://code.google.com/p/pychess/issues/detail?id=821#c9)
- changed status to: Reopened

**Comment by pychess** [comment]
Original [comment 10.](https://code.google.com/p/pychess/issues/detail?id=821#c10) posted by [mattgatto](https://code.google.com/u/mattgatto/) 2014-03-02

This issue was closed by revision 4b9d2a1c38e2.
- changed status to: Fixed
